torch/csrc/api/include/torch/data/dataloader.h

```diff
@@ -298,9 +287,13 @@ template <
 std::unique_ptr<DataLoader<Dataset, Sampler>> make_data_loader(
     Dataset dataset,
     DataLoaderOptions options = DataLoaderOptions()) {
-  const auto size = dataset.size();
-  return torch::make_unique<DataLoader<Dataset, Sampler>>(
-      std::move(dataset), std::move(options), Sampler(size));
+  const optional<size_t> size = dataset.size();
+  AT_CHECK(
+      size.has_value(),
+      "Expected the dataset to be sized in "
+      "order to construct the Sampler");
```

Comment:
The function has a SFINAE guard to only be called when the sampler is constructible from `size_t`, which is assumed to mean it is constructible from the size of the dataset. I should maybe add a separate base class for this instead, in case `size_t` is used for something else